### PR TITLE
fix dropout problem in predict process

### PR DESCRIPTION
**Pull request by weizhenzhao**
image
image

already had it tested , the score is stable

Thanks
Weizhen

### CODE REVIEW

File: rasa_nlu_gao/classifiers/embedding_bert_intent_classifier.py [hunk 1]
```diff
@@ -157,7 +157,8 @@ def __init__(self,
                  graph=None,  # type: Optional[tf.Graph]
                  message_placeholder=None,  # type: Optional[tf.Tensor]
                  intent_placeholder=None,  # type: Optional[tf.Tensor]
-                 y_predict=None   # type: Optional[tf.Tensor]
+                 y_predict=None,   # type: Optional[tf.Tensor]
+                 drop_out=None  # type: Optional[tf.float32]
```
Comment (GaoQ1) [inline rasa_nlu_gao/classifiers/embedding_bert_intent_classifier.py]: 这里没必要给drop_out定义吧，在`self._load_regularization_params()`已经初始化了`self.droprate`，这个值就是`dropout`的值，你这里给了None的话，并不能从外部传值过去。

File: rasa_nlu_gao/classifiers/embedding_bert_intent_classifier.py [hunk 2]
```diff
@@ -192,6 +193,7 @@ def __init__(self,
         self.a_in = message_placeholder
         self.b_in = intent_placeholder
         self.y_predict = y_predict
+        self.drop_out = drop_out
```
Comment (GaoQ1) [inline rasa_nlu_gao/classifiers/embedding_bert_intent_classifier.py]: 同理，这里没必要给定`self.drop_out = drop_out`

File: rasa_nlu_gao/classifiers/embedding_bert_intent_classifier.py [hunk 3]
```diff
@@ -296,8 +299,10 @@ def train(self, training_data, cfg=None, **kwargs):
 
             is_training = tf.placeholder_with_default(False, shape=())
 
+            self.drop_out = tf.placeholder(tf.float32, (), name='drop_out')
+
             # Create a graph for training
-            logits_train = conv_net(self.a_in, num_classes, self.num_hidden_layers, self.hidden_layer_size, self.C2, self.droprate, is_training=True)
+            logits_train = conv_net(self.a_in, num_classes, self.num_hidden_layers, self.hidden_layer_size, self.C2, self.drop_out, is_training=True)
```
Comment (GaoQ1) [inline rasa_nlu_gao/classifiers/embedding_bert_intent_classifier.py]: 这里`self.droprate`也可以吧

**Comment by weizhenzhao** [comment]
> dropout定义的原因

Hi dear

the reason for define drop_out place_holder is in the following link:
[https://stackoverflow.com/questions/38888120/tensorflow-loaded-model-gives-different-predictions](url)

because the drop_out_rate you define previously is just a float number , not a placeholder
so during the graph build process you use this float number to construct your model .
when you execute tf.saver.save method it serialize the model to the disk , also the drop_out rate number
when you  tf.session.restore  the model in the prediction process , 
it also use the drop out rate to predict , 
so it's the drop out rate that cause the predict score unstable .
so here I build a placeholder drop_out to receive your drop out rate number .
and in the prediction process I feed the drop_out placeholder with 0 to ensure the stable of prediction


Thanks
weizhen

File: rasa_nlu_gao/classifiers/embedding_bert_intent_classifier.py (hunk 1 above)
Comment (weizhenzhao) [reply rasa_nlu_gao/classifiers/embedding_bert_intent_classifier.py]: 这个drop_out 指的是一个placeholder  。
`self.drop_out = tf.placeholder(tf.float32,(), name='drop_out')`

来承接您在上边default里面定义的drop_out rate

File: rasa_nlu_gao/classifiers/embedding_bert_intent_classifier.py (hunk 2 above)
Comment (weizhenzhao) [reply rasa_nlu_gao/classifiers/embedding_bert_intent_classifier.py]: 这里也是一个placeholder 然后placeholder的类型是tf.float32 。 shape 是()

File: rasa_nlu_gao/classifiers/embedding_bert_intent_classifier.py (hunk 3 above)
Comment (weizhenzhao) [reply rasa_nlu_gao/classifiers/embedding_bert_intent_classifier.py]: 这块传入一个placeholder 去构建计算图结构，传入droprate是一个数字是非常死的一种做法。就是这种做法导致在预测的时候没办法修改droprate的值。导致预测的Score不稳定

**Comment by GaoQ1** [comment]
> > dropout定义的原因
> 
> Hi dear
> 
> the reason for define drop_out place_holder is in the following link:
> 
> [https://stackoverflow.com/questions/38888120/tensorflow-loaded-model-gives-different-predictions](url)
> 
> because the drop_out_rate you define previously is just a float number , not a placeholder
> so during the graph build process you use this float number to construct your model .
> when you execute tf.saver.save method it serialize the model to the disk , also the drop_out rate number
> when you tf.session.restore the model in the prediction process ,
> it also use the drop out rate to predict ,
> so it's the drop out rate that cause the predict score unstable .
> so here I build a placeholder drop_out to receive your drop out rate number .
> and in the prediction process I feed the drop_out placeholder with 0 to ensure the stable of prediction
> 
> Thanks
> weizhen

OK. I got it.
